### PR TITLE
Improve binance historic OHLCV download speed

### PR DESCRIPTION
**Pull request by xzmeng**
## Summary

This PR attempts to improve historic OHLCV download speed by downloading data from https://data.binance.vision

## Quick changelog
- Override get_historic_ohlcv and implement the logic
- Move the new pair logic from _async_get_historic_ohlcv to get_historic_ohlcv
- Add a new module `freqtrade.exchange.binance_public_data`
## What's new?
1m spot OHLCV 6 years from 65s to 15s - about 4x speedup
1m futures OHLCV 4 years from 10m to 12s - about 50x speedup

### CODE REVIEW

**Comment by xmatthias** [comment]
I'll be quite frank - i'm not sure this is a good idea.

While i apreciate the effort - for such a big PR - i'd apreciate a discussion about it first to avoid wasted time.
It's a TON of code for just one exchange - and for very little gain at that if you think about it carefully.
What do gains look like at more common timeframes (5m, 1h, 1d)? probably very little to none - otherwise these wouldn't be excluded from metrics.

It's also going against a (pretty much undocumented - as far as i could find) page at binance - is there documentation for this page? does binance guarantee to not change the structure? 

also, every update to this endpoing will break downloading 1m and 5m data for users completely ...
Will you handle eventually necessary updates to this in a timely manner (for how long into the future? what response time?)? (don't take this wrongly - it's more a question to edge about "do we really need this")

In it's current state it's adding 800+ lines (to be fair, most of it is tests) for _a little_ more speed - but critically - not speed users will feel day to day (then it might be worth a consideration) - speed in a step which every user will really only have to perform once if they do it properly (freqtrade does incremental downloads - if you need 1m data - run it daily and you'll have an increasing stash of data).

By accepting this - we'd essenitally accept responsibility for maintaining this - which quite honestly - i currently don't see much reason as it's covering an edge-case (1m _initial_ datadownload) - and purely quantity of code - is pretty likely to break in the future - which will make it pretty maintenance intensive.

---

Personally, i think this would be better suited in ccxt - they handle exchange-level communications for us already - and in reality - this is just an API - though i'm not sure they'd consider adding that much code for such an edge-case (start with an issue there first feeling out their willingness if you think about this approach).
I'd not mind calling a different endpoint _in ccxt_ for binance - but having the full logic in freqtrade is not something that excites me.

So at this point - i'm not sure to accept this - but - convince me otherwise :laughing:

File: freqtrade/exchange/exchange.py [hunk 1]
```diff
@@ -2246,7 +2246,7 @@ def get_historic_ohlcv(
                 candle_type=candle_type,
             )
         )
-        logger.info(f"Downloaded data for {pair} with length {len(data)}.")
+        logger.info(f"Downloaded data for {pair} from ccxt with length {len(data)}.")
```
Comment (xmatthias) [inline freqtrade/exchange/exchange.py]: a user won't care about this.

File: freqtrade/exchange/binance_public_data.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    dfs: list[DataFrame | None] = []
+
+    connector = aiohttp.TCPConnector(limit=100)
+    async with aiohttp.ClientSession(connector=connector) as session:
+        coroutines = [
+            get_daily_ohlcv(asset_type, symbol, timeframe, date, session)
+            for date in date_range(start, end)
+        ]
+        # the HTTP connections has been throttled by TCPConnector
+        for batch in itertools.batched(coroutines, 1000):
```
Comment (xmatthias) [inline freqtrade/exchange/binance_public_data.py]: itertools.batched was added in 3.12 ... we support 3.10 and up - so you can't use this (there's a helper - `chunks` - in misc that essentially does the same thing).

File: freqtrade/exchange/binance_public_data.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    e.g. BTC/USDT -> BTCUSDT, BTC/USDT:USDT -> BTCUSDT
+    """
+    if ":" in symbol:
+        parts = symbol.split(":")
+        if len(parts) != 2:
+            raise ValueError(f"Cannot recognize symbol: {symbol}")
+        return parts[0].replace("/", "")
+    else:
+        return symbol.replace("/", "")
+
```
Comment (xmatthias) [inline freqtrade/exchange/binance_public_data.py]: I'm not sure that's necessarily correct for all cases - better rely on the ccxt markets object - it contains the original symbol.

File: freqtrade/exchange/binance.py [hunk 4]
```diff
@@ -98,38 +102,72 @@ def additional_exchange_init(self) -> None:
...
                 )
+                if until_ms and since_ms >= until_ms:
+                    logger.warning(
+                        f"No available candle-data for {pair} before"
+                        f"{dt_from_ts(until_ms).isoformat()}"
+                    )
+                    return DataFrame(columns=DEFAULT_DATAFRAME_COLUMNS)
 
-        return await super()._async_get_historic_ohlcv(
+        if timeframe in ["1m", "5m"] and candle_type in [CandleType.SPOT, CandleType.FUTURES]:
```
Comment (xmatthias) [inline freqtrade/exchange/binance.py]: can you performance test this with 1s? (also, why is futures downloads so much slower than spot in your comparison?)

download-data (and --timeframe-detail) do support 1s ... but downloading that data takes ages.

**Comment by froggleston** [comment]
What might be a good outcome of this is some documentation or implementation for trade data as this currently is incredibly slow from the API. Would this endpoint be useful to mention to people who want to use public trade data?

**Comment by xmatthias** [comment]
> Would this endpoint be useful to mention to people who want to use public trade data?

from what i gather - only for backtesting. Data is not uptodate to "today" - so for a live run, we'd always have to fall back to the exchange API.

It's not directly an endpoint either - more like a web directory you can browse through :laughing: (you can just navigate to the base directory and start "browsing", really)

**Comment by xzmeng** [comment]
I hesitated to submit this PR myself because the OHLCV download speed for spot markets is already quite fast. However, for futures, the rate limit is significantly lower—2400 weights per minute with a weight of 5 per 1000 klines—about one-sixth the rate of spot (6000 weights per minute with a weight of 2 per 1000 klines). This discrepancy makes the download speed for futures much slower compared to spot. I've prepared a comparison table, and the original data can be found [here](https://github.com/freqtrade/freqtrade/blob/45bf046645bb20a57451327606b055ca6207c6fc/tests/exchange/binance_ohlcv_compare.py) 
|branch|trading_mode|pair|timeframe|timerange|duration|download seconds|comparison|
|-------|--------------|----|----------|----------|--------|---------------|------------|
|pr|spot|ETH/USDT|1s|20240101-20240501|4 months|28.2|1x|
|develop|spot|ETH/USDT|1s|20240101-20240501|4 months|257.9|9x|
|pr|spot|ETH/USDT|1m|20180101-20240101|6 years|17.7|1x|
|develop|spot|ETH/USDT|1m|20180101-20240101|6 years|86.4|5x|
|pr|spot|ETH/USDT|5m|20180101-20240101|6 years|11.3|1x|
|develop|spot|ETH/USDT|5m|20180101-20240101|6 years|20.1|2x|
|pr|futures|ETH/USDT:USDT|1m|20200101-20240101|4 years|15.9|1x|
|develop|futures|ETH/USDT:USDT|1m|20200101-20240101|4 years|537.5|34x|
|pr|futures|ETH/USDT:USDT|5m|20200101-20240101|4 years|10.1|1x|
|develop|futures|ETH/USDT:USDT|5m|20200101-20240101|4 years|111.13|11x|

The **download seconds** is the time elapsed by running `freqtrade download-data` from shell.

File: freqtrade/exchange/binance.py (hunk 4 above)
Comment (xzmeng) [reply freqtrade/exchange/binance.py]: have posted a table for comparison and explained the reason why futures is much slower.

File: freqtrade/exchange/exchange.py (hunk 1 above)
Comment (xzmeng) [reply freqtrade/exchange/exchange.py]: Yes, but leave it as it is, you will always see "with length 0" if  the timerange does not include the latest 2 days. So I add another message and changed this one. Better way may be need to refactor the current code.

**Comment by xzmeng** [comment]
> What might be a good outcome of this is some documentation or implementation for trade data as this currently is incredibly slow from the API. Would this endpoint be useful to mention to people who want to use public trade data?

It can be slow even from archive files, I saw DOGE once have 2GB aggregated trades one day.

I have not played freqtrade with trades data so far, so I can't answer the question.

**Comment by xzmeng** [comment]
> It's also going against a (pretty much undocumented - as far as i could find) page at binance - is there documentation for this page? does binance guarantee to not change the structure?

take a look (if you haven't) at https://github.com/binance/binance-public-data
The data structure is documented on github and I have used it more than 3 years, it's quite stable IMO (if binance won't collapse or bankrupt, haha).

> also, every update to this endpoing will break downloading 1m and 5m data for users completely ...

As above said, the page is not likely often changed. And we can(or must) fall back to rest API if it's unavailable.

>In it's current state it's adding 800+ lines (to be fair, most of it is tests)

Maybe I can clean the test, I am really not familiar with the unittest, my poorly desgined tests might make it overly bloated,  but the core logic is just about a hundred of lines.

>Personally, i think this would be better suited in ccxt

I don't think ccxt would like to accept this 😅

>for such a big PR

No, the tests bloated it, it's indeed a minor improvement (what might be not much useful, but faster speed can make me happy, I believe there be many "light" users of freqtrade who do not maintain an ohlcv database, and just download as they need, I it can lead to better user experience when the download speed can be faster -- currently binance is the mostly used exchange). finally, accept it or not, I completely understand your considerations, consistency and maintainability count.  😆

File: freqtrade/exchange/binance.py [hunk 5]
```diff
@@ -98,38 +102,72 @@ def additional_exchange_init(self) -> None:
...
+                    logger.warning(
+                        f"No available candle-data for {pair} before"
+                        f"{dt_from_ts(until_ms).isoformat()}"
+                    )
+                    return DataFrame(columns=DEFAULT_DATAFRAME_COLUMNS)
 
-        return await super()._async_get_historic_ohlcv(
+        if timeframe in ["1s", "1m", "5m"] and candle_type in [CandleType.SPOT, CandleType.FUTURES]:
+            df = self.loop.run_until_complete(
+                binance_public_data.fetch_ohlcv(
```
Comment (xmatthias) [inline freqtrade/exchange/binance.py]: i would rename the function (it's not fetching - it really is doing a whole download and a lot more) - and import it as function instead of the whole module ...

File: freqtrade/exchange/binance_public_data.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+        return None if no data available in the time range
+    """
+    if candle_type == CandleType.SPOT:
+        asset_type = "spot"
+    elif candle_type == CandleType.FUTURES:
+        asset_type = "futures/um"
+    else:
+        raise ValueError(f"Unsupported CandleType: {candle_type}")
+    symbol = symbol_ccxt_to_binance(pair)
+    start = dt_from_ts(since_ms)
```
Comment (xmatthias) [inline freqtrade/exchange/binance_public_data.py]: I think starting date handling is a key part - and it's currently unclear to me how you're handling the following (not necessarily tied to this line - but most likely close to it)

Assume i have existing data for a pair - which ends on 2024-10-25 at 11:35 am.
What will be the first file to be downloaded? 
How do you ensure the data ain't stored twice or contains holes (like - the rest of the 25th is missing completely)?
The same actually applies at the end of the timerange (where we assume "data stops here - let's use api from here onwards).

File: freqtrade/exchange/binance_public_data.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+def concat(dfs) -> DataFrame:
+    if all(df is None for df in dfs):
+        return DataFrame()
+    else:
+        return pd.concat(dfs)
+
+
+async def _fetch_ohlcv(
+    asset_type: str,
```
Comment (xmatthias) [inline freqtrade/exchange/binance_public_data.py]: You might want to improve the naming of this variable
It suggests "spot" / "futures" - but actually contains the directory we intend to download ... which is `futures/um` ... so it's more like "asset_type_url_part - or something similar ?

File: freqtrade/exchange/binance_public_data.py (hunk 6 above)
Comment (xzmeng) [reply freqtrade/exchange/binance_public_data.py]: >What will be the first file to be downloaded?

2024-10-25
https://github.com/freqtrade/freqtrade/blob/3c76af9dab59973e41f6cbde31356c3dfe4f56b5/freqtrade/exchange/binance_public_data.py#L125-L129
>How do you ensure the data ain't stored twice

I don't. The logic should be implemented by the caller of `get_historic_ohlcv`, so I just assumed If I return the  data of  timerange [since, until) (a left-closed right open timerange), it will work fine (I know the assumption might be wrong, ~... and it is~, but it seems works this way)

https://github.com/freqtrade/freqtrade/blob/3c76af9dab59973e41f6cbde31356c3dfe4f56b5/freqtrade/exchange/binance_public_data.py#L65-L68

>or contains holes (like - the rest of the 25th is missing completely)?

`fetch_ohlcv` has two modes with different assumptions:
1. `stop_on_404` is `False`: This mode we trust https://data.binance.vision provide the same data as rest API, i.e. a 404 error means there is no data in that day, so no holes.
2. `stop_on_404` is `True`: This mode we don't trust https://data.binance.vision, any 404 error means we should stop the following data download, and should fall back to rest API, so no holes.

https://github.com/freqtrade/freqtrade/blob/3c76af9dab59973e41f6cbde31356c3dfe4f56b5/freqtrade/exchange/binance_public_data.py#L111-L121

>The same actually applies at the end of the timerange (where we assume "data stops here - let's use api from here onwards)

https://github.com/freqtrade/freqtrade/blob/3c76af9dab59973e41f6cbde31356c3dfe4f56b5/freqtrade/exchange/binance.py#L148-L163

rest means REST API or rest ...

----------
~I think this kind of corner cases should be tested by some real command line rather than plenty of mocks, if you think this PR worth to proceed on, I'd like to add some test cases.~ Forget the silly words ...

File: freqtrade/exchange/binance_public_data.py (hunk 7 above)
Comment (xzmeng) [reply freqtrade/exchange/binance_public_data.py]: Naming is always the hardest part ......

File: freqtrade/exchange/binance_public_data.py (hunk 3 above)
Comment (xzmeng) [reply freqtrade/exchange/binance_public_data.py]: The reason why I'm firstly reluctant to use ccxt is that exchangeInfo endpoint consumes a lot of time (maybe 2 seconds) and I need to patch this function or ccxt in the unittest.

I have checked all USD-M futures and I can confirm that all of the **perpetual contracts** (but not include dilivery) conform to this pattern: `COIN/USDT:USDT` or `COIN/USDC:USDC` (although ccxt and binance may use different COIN names)

File: freqtrade/exchange/exchange.py (hunk 1 above)
Comment (xzmeng) [reply freqtrade/exchange/exchange.py]: moved the message to `get_historic_ohlcv` to its single caller
https://github.com/freqtrade/freqtrade/pull/10858/commits/cf0f232635bc14be9e977646f8b0744c25cbc7e0

File: freqtrade/exchange/binance_public_data.py (hunk 3 above)
Comment (xmatthias) [reply freqtrade/exchange/binance_public_data.py]: > (although ccxt and binance may use different COIN names)

wouldn't this mean that ccxt uses BTC/USDT:USDT - but on bybit it's XBT_USDT ... ? which would mean the logic above will fail.
Best example is `YOYOW/BTC` - which is actually `YOYOBTC` (missing a W) at the exchange level - so i don't think you can get around using the ccxt market object.

While it'll add 2s to the overall time (it really doesn't, you already initialized the exchange to get to this function) - it'll also improve reliability - and you can really pass that in as object (`exchange.markets` - so you don't need the exchange object itself).

File: freqtrade/exchange/binance_public_data.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    symbol: str,
+    timeframe: str,
+    start: datetime.date,
+    end: datetime.date,
+    stop_on_404: bool,
+) -> DataFrame:
+    dfs: list[DataFrame | None] = []
+
+    connector = aiohttp.TCPConnector(limit=100)
+    async with aiohttp.ClientSession(connector=connector) as session:
```
Comment (xmatthias) [inline freqtrade/exchange/binance_public_data.py]: How will this handle proxies? 
in dream mode - it'd take it from the ccxt_config.httpsProxy object (https://www.freqtrade.io/en/latest/configuration/#using-a-proxy-with-freqtrade) - but i'm not sure if that's a good idea as this can really be httpProxy, socksProxy, (ccxt documentation has them all).

the alternative is to use the env variables ... maybe that's builtin - but i didn't really check for this.

File: freqtrade/exchange/binance.py [hunk 9]
```diff
@@ -98,38 +102,74 @@ def additional_exchange_init(self) -> None:
...
         until_ms: Optional[int] = None,
-    ) -> OHLCVResponse:
+    ) -> DataFrame:
         """
         Overwrite to introduce "fast new pair" functionality by detecting the pair's listing date
         Does not work for other exchanges, which don't return the earliest data when called with "0"
         :param candle_type: Any of the enum CandleType (must match trading mode!)
         """
         if is_new_pair:
-            x = await self._async_get_candle_history(pair, timeframe, candle_type, 0)
```
Comment (xmatthias) [inline freqtrade/exchange/binance.py:116]: I think we'll want to have the new mode optional (but on by default!).

Implicitly, this makes the previous override of "get the oldest available data" necessary again.
As a side effect - it'll also make removing this mode simple should that be necessary in the future - as it'll get a pretty clear boundary (i don't expect to need to do this - but who knows what problems can arise - from unreliable downloads, to disappearing pages, ... :laughing: )

I'm not entirely sure how availability of the binance.vision page is across the world - so by forcing this onto users, we may break data-download for them completely if their location isn't able to access binance.vision.

File: tests/exchange/binance_ohlcv_compare.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
+"""
+This file is meant to test the OHLCV download performance between
+from rest API and from data.binance.vision
```
Comment (xmatthias) [inline tests/exchange/binance_ohlcv_compare.py]: do we intend to keep this file? if yes, then maybe it should be moved to scripts.

I guess it's however only here for performance testing ...?

File: freqtrade/exchange/binance.py (hunk 9 above)
Comment (xmatthias) [reply freqtrade/exchange/binance.py:116]: As a side note - is this (get the oldest possible datapoint) handled (or even necessary) in the new mode? like - if i download with 20170101- - but the pair was listed in 2023 ... i'd be making 1000nds of calls to get daily files ... while in reality - i'd only need 300 ...

**Review by xmatthias — COMMENTED** [review]
Let's push on with this and get this to "mergable" ;)

I did add a few comments - though i didn't look at every detail so far.

File: freqtrade/exchange/binance_public_data.py (hunk 3 above)
Comment (xzmeng) [reply freqtrade/exchange/binance_public_data.py]: Agreed we should use CCXT.

> wouldn't this mean that ccxt uses BTC/USDT:USDT - but on bybit it's XBT_USDT ... ? which would mean the logic above will fail.

the function is only for Binance

File: freqtrade/exchange/binance.py (hunk 9 above)
Comment (xzmeng) [reply freqtrade/exchange/binance.py:116]: >Implicitly, this makes the previous override of "get the oldest available data" necessary again.

I did not remove this part, but moved it to `get_historic_ohlcv` so that both modes can utilize it.

>is this (get the oldest possible datapoint) handled (or even necessary) in the new mode?

Yes, as said above.

File: freqtrade/exchange/binance_public_data.py (hunk 8 above)
Comment (xzmeng) [reply freqtrade/exchange/binance_public_data.py]: Using the environment variables seems more correct, however, it might be confusing for users why we need two proxies for download binance candlesticks 😅

Currently use  enviroment variables:
[feat: support proxy from environment variables](https://github.com/freqtrade/freqtrade/pull/10858/commits/d7555e1f2971b6bcbbd4d978bd0a81e436c91f3d)

File: tests/exchange/binance_ohlcv_compare.py (hunk 10 above)
Comment (xzmeng) [reply tests/exchange/binance_ohlcv_compare.py]: Removed

File: freqtrade/exchange/binance.py (hunk 9 above)
Comment (xzmeng) [reply freqtrade/exchange/binance.py:116]: - Added a paramter `only_from_ccxt` in `Exchange.get_historic_ohlcv` which defaults to `False`, we can pass `True` to disable the new mode.
- Stripped the new mode into an isolated method `Binance.get_historic_ohlcv_fast`, the new mode can be easily eliminated when we need.

[refactor: strip fast download logic into a method](https://github.com/freqtrade/freqtrade/pull/10858/commits/acc53065e5fa7ab5197073276306dc9dc3adbfa3)

File: freqtrade/exchange/binance_public_data.py (hunk 3 above)
Comment (xzmeng) [reply freqtrade/exchange/binance_public_data.py]: [refactor: use CCXT for pair to symbol conversion](https://github.com/freqtrade/freqtrade/pull/10858/commits/37726fba58834e922bc8079141aa57d5585f08ec)

File: freqtrade/exchange/binance.py (hunk 5 above)
Comment (xzmeng) [reply freqtrade/exchange/binance.py]: From a caller's perspective, it just returns the ohlcv requested. Of course `fetch_ohlcv` can be confused with some method in CCXT so I imported the module name for readability.

How about `download_archive_ohlcv`(sounds not good ...)?  I'd like to hear your advice for the naming.

Comment (xmatthias) [reply freqtrade/exchange/binance.py]: `download_archive_ohlcv` sounds pretty good, actually - and explains pretty well what it does

**Comment by xmatthias** [comment]
Something seems off  ... not sure what exactly just yet
It seems to fall back "way" too early to exchange calls ... so for XRP ... this seems a bit odd.

with `freqtrade download-data --trading-mode futures --exchange binance --pairs XRP/USDT:USDT -t 1m  --timerange 20170101- -v` - i get the following: 


```
2024-11-13 18:18:54,244 - freqtrade.exchange.binance_public_data - DEBUG - Successfully downloaded                                         
https://data.binance.vision/data/futures/um/daily/klines/XRPUSDT/1m/XRPUSDT-1m-2022-10-01.zip                                              
2024-11-13 18:18:54,280 - freqtrade.exchange.binance_public_data - DEBUG - Successfully downloaded                                         
https://data.binance.vision/data/futures/um/daily/klines/XRPUSDT/1m/XRPUSDT-1m-2022-08-12.zip                                              
2024-11-13 18:18:54,380 - freqtrade.exchange.binance_public_data - DEBUG - Successfully downloaded                                         
https://data.binance.vision/data/futures/um/daily/klines/XRPUSDT/1m/XRPUSDT-1m-2022-08-20.zip                                              
2024-11-13 18:18:54,525 - freqtrade.exchange.binance_public_data - DEBUG - Successfully downloaded                                         
https://data.binance.vision/data/futures/um/daily/klines/XRPUSDT/1m/XRPUSDT-1m-2022-09-02.zip                                              
2024-11-13 18:18:54,961 - freqtrade.exchange.binance_public_data - DEBUG - Successfully downloaded                                         
https://data.binance.vision/data/futures/um/daily/klines/XRPUSDT/1m/XRPUSDT-1m-2022-09-27.zip                                              
2024-11-13 18:18:54,966 - freqtrade.exchange.binance_public_data - DEBUG - Abort downloading from data.binance.vision due to 404, 330      <==== i added this 330 - which corresponds to current_day
2024-11-13 18:18:54,998 - freqtrade.exchange.binance_public_data - DEBUG - Downloaded data for XRP/USDT:USDT from                          
https://data.binance.vision with length 473259.                                                                                            
2024-11-13 18:18:55,022 - freqtrade.exchange.exchange - DEBUG - one_call: 60000000 msecs (16 hours ago)                                    
2024-11-13 18:18:55,028 - freqtrade.exchange.exchange - DEBUG - Fetching pair XRP/USDT:USDT, futures, interval 1m, since 1606694400000     
(2020-11-30T00:00:00+00:00) ...                                                                                                            
2024-11-13 18:18:55,030 - freqtrade.exchange.exchange - DEBUG - Fetching pair XRP/USDT:USDT, futures, interval 1m, since 1606754400000     
(2020-11-30T16:40:00+00:00) ...                                                                                                            
2024-11-13 18:18:55,032 - freqtrade.exchange.exchange - DEBUG - Fetching pair XRP/USDT:USDT, futures, interval 1m, since 1606814400000     
(2020-12-01T09:20:00+00:00) ...                                                                                                            
2024-11-13 18:18:55,034 - freqtrade.exchange.exchange - DEBUG - Fetching pair XRP/USDT:USDT, futures, interval 1m, since 1606874400000  
```

File: freqtrade/exchange/binance_public_data.py [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+                        if current_day == 1:
+                            logger.warning(
+                                "Failed to use fast download, fall back to rest API download, this "
+                                "can take more time. If you're downloading BTC/USDT:USDT, "
+                                "ETH/USDT:USDT, BCH/USDT:USDT, please first download "
+                                "data before 2020 (using `--timerange yyyymmdd-20200101`), and "
+                                "then download the full data you need."
+                            )
+                        logger.debug("Abort downloading from data.binance.vision due to 404")
+                        return concat(dfs)
```
Comment (xmatthias) [inline freqtrade/exchange/binance_public_data.py]: This message should definitely be higher than debug level.
It's rather important for us to diagnose "why" something it's taking forever

Comment (xzmeng) [reply freqtrade/exchange/binance_public_data.py]: 截屏2024-11-14 03 11 33

There is a missing week of data. `asyncio.gather` waited all tasks to complete, which resulted unnecessary data download before falling back to API , should we reduce the chunk size from 1000 to a smaller one? (100?)

Comment (951214) [reply freqtrade/exchange/binance_public_data.py:162]: Resovled the unnecessary download by replace `asyncio.gather` with `asyncio.create_task` and cancel the remaining tasks when we return:
[feat: cancel uncompleted tasks before return](https://github.com/freqtrade/freqtrade/pull/10858/commits/6b18c4f24c16213c52ec098863b787298322072e)

Added some user friendly warning messages when falling back to rest API:
[chore: add user friendly warnings](https://github.com/freqtrade/freqtrade/pull/10858/commits/8baa0f73103d86057195fa1a4dac74b8b30bd6e4)

Comment (xzmeng) [reply freqtrade/exchange/binance_public_data.py]: Resovled the unnecessary download by replace `asyncio.gather` with `asyncio.create_task` and cancel the remaining tasks when we return:
[feat: cancel uncompleted tasks before return](https://github.com/freqtrade/freqtrade/pull/10858/commits/6b18c4f24c16213c52ec098863b787298322072e)

Added some user friendly warning messages when falling back to rest API:
[chore: add user friendly warnings](https://github.com/freqtrade/freqtrade/pull/10858/commits/8baa0f73103d86057195fa1a4dac74b8b30bd6e4)

File: freqtrade/exchange/binance.py (hunk 5 above)
Comment (xzmeng) [reply freqtrade/exchange/binance.py]: Renamed

File: freqtrade/exchange/binance_public_data.py (hunk 7 above)
Comment (xzmeng) [reply freqtrade/exchange/binance_public_data.py]: Renamed to `asset_type_url_segment`

**Comment by xmatthias** [comment]
@xzmeng is this actually still a wip - or is it actually ready from your point of view but waiting my (late) review? :sweat_smile:

**Comment by xzmeng** [comment]
it's ready for reviewing~

File: freqtrade/exchange/binance.py [hunk 12]
```diff
@@ -97,41 +102,114 @@ def additional_exchange_init(self) -> None:
         except ccxt.BaseError as e:
             raise OperationalException(e) from e
 
-    async def _async_get_historic_ohlcv(
+    def get_historic_ohlcv(
```
Comment (xmatthias) [inline freqtrade/exchange/binance.py:105]: Is there a particular reason you didn't remain in the async context, but moved this "up" from that? 
in the end, we'll still need the (same) async context for both download methods - just that now we need "run_until_complete" to do actual work :D 

(for clarity: it's not necessarily a request for change - more an asking for reason)

File: tests/exchange_online/test_binance_compare_ohlcv.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+            symbols = await self.get_symbols()
+            await asyncio.gather(*[self.check_one_symbol(symbol) for symbol in symbols])
+            return self.mismatch
+
+
+@pytest.mark.skipif(
+    not bool(os.environ.get("TEST_BINANCE_COMPARE_OHLCV")),
+    reason="Simply to demonstrate the availabity of the archive endpoint",
+)
+async def test_binance_compare_ohlcv():
```
Comment (xmatthias) [inline tests/exchange_online/test_binance_compare_ohlcv.py]: I'm torn on this whole file (if it's worth keeping, really).

it seems pretty useful per se - but it won't run - and is prone to fail (it currently fails on 3 new pairs).
On the other hand - it only tests "is the pair available" - not "is the full timerange available" - which is where i suspect more problems / need for future investigation.

File: freqtrade/exchange/binance_public_data.py [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+            asset_type_url_segment = "futures/um"
+        else:
+            raise ValueError(f"Unsupported CandleType: {candle_type}")
+
+        if markets:
+            symbol = markets[pair]["id"]
+        else:
+            binance = ccxt.binance()
+            binance.load_markets()
+            symbol = binance.markets[pair]["id"]
```
Comment (xmatthias) [inline freqtrade/exchange/binance_public_data.py]: Why would this be necessary? it seems to be a pointless fallback. 
we control all call paths to this method (i don't see this as public api users would call directly) - and markets is guaranteed to be loaded at the point you're reaching here.

if it's not loaded - you'd never get to this point anyway as no pairs would be seen as valid.

**Review by xmatthias — COMMENTED** [review]
I did a few minor edits - mostly around naming to avoid naming conflicts / confusion.

There's also a few questions that popped up - but the rest seems pretty much great :+1:

File: freqtrade/exchange/binance_public_data.py (hunk 14 above)
Comment (xzmeng) [reply freqtrade/exchange/binance_public_data.py]: yes it's not necessary.

File: freqtrade/exchange/binance.py (hunk 12 above)
Comment (xzmeng) [reply freqtrade/exchange/binance.py:105]: the async method returns an OHLCVRespnse which is kind of raw data, returning a dataframe is helpful for dev and makes the data flow concise.

File: tests/exchange_online/test_binance_compare_ohlcv.py (hunk 13 above)
Comment (xzmeng) [reply tests/exchange_online/test_binance_compare_ohlcv.py]: this file is like the previously removed one, it gives us some confidence that not too many pairs will fail at first.

I think we can remove it.

Or add it to CI, last time when I run it there are 3 futures pairs failed: BTC ETH BCH, it is deliberately to fail to see the detail. and these 3 pairs are listed quite early before 2020, so 
I guess the result is relatively stable.

File: freqtrade/exchange/binance.py (hunk 12 above)
Comment (xmatthias) [reply freqtrade/exchange/binance.py:105]: works for me ;) as said - just a confirming question

File: tests/exchange_online/test_binance_compare_ohlcv.py (hunk 13 above)
Comment (xmatthias) [reply tests/exchange_online/test_binance_compare_ohlcv.py]: > I guess the result is relatively stable.

it's not unfortunately - not without further filters
I've ran it yesterday evening, and it failed on 3 more pairs (6 in total) - all pairs (not in your list above) that have only been listed within the last day or two.

As binance lists pairs all the time - it'd be a test we expect to fail one day a week at least ... if not more 😆

Comment (xzmeng) [reply tests/exchange_online/test_binance_compare_ohlcv.py]: it's expected because the archive upload always have about 30 hours delay. let's just remove it~

**Review by xmatthias — APPROVED** [review]
I'd say let's give this a go.

We'll see if others encounter problems - i intentionally waited until the release was cut so this get's the maximum time in "develop only".

LGTM :+1:
